File: js/sdk-sui/src/types/game.ts
import { bcs } from '@mysten/bcs'
import { Address, Parser } from './parser'
import {
    Bonus,
    PlayerBalance,
    CheckpointOnChain,
    EntryLock,
    EntryType,
    GameAccount,
    PlayerJoin,
    PlayerDeposit,
    ServerJoin,
    VoteType,
    ENTRY_LOCKS,
    VOTE_TYPES,
    DEPOSIT_STATUS,
} from '@race-foundation/sdk-core'

const BonusSchema = bcs.struct('BonusSchema', {
    id: Address,
    identifier: bcs.string(),
    tokenAddr: bcs.string(),
    amount: bcs.u64(),
})

const BonusParser: Parser<Bonus, typeof BonusSchema> = {
    schema: BonusSchema,
    transform: (input: typeof BonusSchema.$inferType): Bonus => {
        return {
            identifier: input.identifier,
            tokenAddr: input.tokenAddr,
            amount: BigInt(input.amount)
        }
    }
}

const PlayerBalanceSchema = bcs.struct('PlayerBalance', {
    playerId: bcs.u64(),
    balance: bcs.u64(),
})

const PlayerBalanceParser: Parser<PlayerBalance, typeof PlayerBalanceSchema> = {
    schema: PlayerBalanceSchema,
    transform: (input: typeof PlayerBalanceSchema.$inferType): PlayerBalance => {
        return {
            playerId: BigInt(input.playerId),
            balance: BigInt(input.balance),
        }
    }
}

const EntryTypeCashSchema = bcs.struct('EntryTypeCash', {
    min_deposit: bcs.u64(), max_deposit: bcs.u64()
})

const EntryTypeTicketSchema = bcs.struct('EntryTypeTicket', {
    amount: bcs.u64()
})

const EntryTypeGatingSchema = bcs.struct('EntryTypeCash', {
    collection: bcs.string()
})

// Define the schema for EntryType
const EntryTypeSchema = bcs.enum('EntryTypeSchema', {
    Cash: EntryTypeCashSchema,
    Ticket: EntryTypeTicketSchema,
    Gating: EntryTypeGatingSchema,
    Disabled: null,
})

const EntryTypeParser: Parser<EntryType, typeof EntryTypeSchema> = {
    schema: EntryTypeSchema,
    transform: (input: typeof EntryTypeSchema.$inferType): EntryType => {
        if (input.$kind === 'Cash') {
            return {
                minDeposit: BigInt(input.Cash.min_deposit),
                maxDeposit: BigInt(input.Cash.max_deposit),
                kind: 'cash',
            }
        } else if (input.$kind === 'Ticket') {
            return {
                amount: BigInt(input.Ticket.amount),
                kind: 'ticket',
            }
        } else if (input.$kind == 'Gating') {
            return {
                collection: input.Gating.collection,
                kind: 'gating',
            }
        } else {
            return { kind: 'disabled' }
        }
    }
}

// Define the schema for Game
const GameSchema = bcs.struct('Game', {
    addr: Address,
    version: bcs.string(),
    title: bcs.string(),
    bundleAddr: Address,
    tokenAddr: bcs.string(),
    ownerAddr: Address,
    recipientAddr: Address,
    transactorAddr: bcs.option(Address),
    accessVersion: bcs.u64(),
    settleVersion: bcs.u64(),
    maxPlayers: bcs.u16(),
    players: bcs.vector(bcs.struct('PlayerJoin', {
        addr: Address,
        position: bcs.u16(),
        accessVersion: bcs.u64(),
        verifyKey: bcs.string(),
    })),
    deposits: bcs.vector(bcs.struct('PlayerDeposit', {
        addr: Address,
        amount: bcs.u64(),
        accessVersion: bcs.u64(),
        settleVersion: bcs.u64(),
        status: bcs.u8(), // This should map to DepositStatus in the transform function
    })),
    servers: bcs.vector(bcs.struct('ServerJoin', {
        addr: Address,
        endpoint: bcs.string(),
        accessVersion: bcs.u64(),
        verifyKey: bcs.string(),
    })),
    balance: bcs.u64(),
    dataLen: bcs.u32(),
    data: bcs.vector(bcs.u8()),
    votes: bcs.vector(bcs.struct('Vote', {
        voter: Address,
        votee: Address,
        voteType: bcs.u8(), // This should map to VoteType in the transform function
    })),
    unlockTime: bcs.option(bcs.u64()),
    entryType: EntryTypeSchema,
    checkpointOnChain: bcs.vector(bcs.u8()),
    entryLock: bcs.u8(),  // This should map to EntryLock in the transform function
    bonuses: bcs.vector(BonusSchema),
    balances: bcs.vector(PlayerBalanceSchema),
})

// Transform function to convert from BCS to the TypeScript type
export const GameAccountParser: Parser<GameAccount, typeof GameSchema> = {
    schema: GameSchema,
    transform: (input: typeof GameSchema.$inferType): GameAccount => {

        console.info('input.checkpointOnChain =>', input.checkpointOnChain)
        console.info('input.transactorAddr =>', input.transactorAddr)

        return {
            addr: input.addr,
            title: input.title,
            bundleAddr: input.bundleAddr,
            tokenAddr: input.tokenAddr,
            ownerAddr: input.ownerAddr,
            settleVersion: BigInt(input.settleVersion),
            accessVersion: BigInt(input.accessVersion),
            players: Array.from(input.players).map((player) => ({
                addr: player.addr,
                position: player.position,
                accessVersion: BigInt(player.accessVersion),
                verifyKey: player.verifyKey,
            })),
            deposits: Array.from(input.deposits).map((deposit) => ({
                addr: deposit.addr,
                amount: BigInt(deposit.amount),
                accessVersion: BigInt(deposit.accessVersion),
                settleVersion: BigInt(deposit.settleVersion),
                status: DEPOSIT_STATUS[deposit.status],
            })),
            servers: Array.from(input.servers).map((server) => ({
                addr: server.addr,
                endpoint: server.endpoint,
                accessVersion: BigInt(server.accessVersion),
                verifyKey: server.verifyKey,
            })),
            transactorAddr: input.transactorAddr ?? undefined,
            votes: Array.from(input.votes).map((vote) => ({
                voter: vote.voter,
                votee: vote.votee,
                voteType: VOTE_TYPES[vote.voteType],
            })),
            unlockTime: input.unlockTime ? BigInt(input.unlockTime) : undefined,
            maxPlayers: input.maxPlayers,
            dataLen: input.dataLen,
            data: Uint8Array.from(input.data),
            entryType: EntryTypeParser.transform(input.entryType),
            recipientAddr: input.recipientAddr,
            checkpointOnChain: input.checkpointOnChain.length > 0
                ? CheckpointOnChain.fromRaw(Uint8Array.from(input.checkpointOnChain))
                : undefined,
            entryLock: ENTRY_LOCKS[input.entryLock],
            bonuses: Array.from(input.bonuses).map((bonus) => (
                BonusParser.transform(bonus))
            ),
            balances: Array.from(input.balances).map((balance) => (
                PlayerBalanceParser.transform(balance)
            ))
        }
    },
}

File: transport/src/solana/types/state/game.rs
use borsh::{BorshDeserialize, BorshSerialize};
use race_core::error::Error;
use race_core::types::DepositStatus;
use race_core::{
    checkpoint::CheckpointOnChain,
    entry_type::EntryType,
    types::{EntryLock, GameAccount, VoteType},
};
use solana_sdk::pubkey::Pubkey;
use crate::solana::types::state::players::PlayerJoin;

#[cfg_attr(test, derive(PartialEq, Eq))]
#[derive(Default, BorshDeserialize, BorshSerialize, Clone, Debug)]
pub struct PlayerDeposit {
    pub addr: Pubkey,
    pub amount: u64,
    pub access_version: u64,
    pub settle_version: u64,
    pub status: DepositStatus,
}

impl From<PlayerDeposit> for race_core::types::PlayerDeposit {
    fn from(value: PlayerDeposit) -> Self {
        Self {
            addr: value.addr.to_string(),
            amount: value.amount,
            access_version: value.access_version,
            settle_version: value.settle_version,
            status: value.status,
        }
    }
}

#[cfg_attr(test, derive(PartialEq, Eq))]
#[derive(Default, BorshDeserialize, BorshSerialize, Clone, Debug)]
pub struct ServerJoin {
    pub addr: Pubkey,
    pub endpoint: String,
    pub access_version: u64,
}

impl From<ServerJoin> for race_core::types::ServerJoin {
    fn from(value: ServerJoin) -> Self {
        Self {
            addr: value.addr.to_string(),
            endpoint: value.endpoint,
            access_version: value.access_version,
        }
    }
}

#[cfg_attr(test, derive(PartialEq, Eq))]
#[derive(BorshDeserialize, BorshSerialize, Clone, Debug)]
pub struct Vote {
    pub voter: Pubkey,
    pub votee: Pubkey,
    pub vote_type: VoteType,
}

#[cfg_attr(test, derive(PartialEq, Eq))]
#[derive(BorshDeserialize, BorshSerialize, Clone, Debug)]
pub struct Bonus {
    pub identifier: String,
    pub stake_addr: Pubkey,
    pub token_addr: Pubkey,
    pub amount: u64,
}

impl From<Bonus> for race_core::types::Bonus {
    fn from(value: Bonus) -> Self {
        Self {
            identifier: value.identifier,
            token_addr: value.token_addr.to_string(),
            amount: value.amount,
        }
    }
}

#[cfg_attr(test, derive(PartialEq, Eq))]
#[derive(BorshDeserialize, BorshSerialize, Clone, Debug)]
pub struct PlayerBalance {
    pub player_id: u64,
    pub balance: u64,
}

impl From<PlayerBalance> for race_core::types::PlayerBalance {
    fn from(value: PlayerBalance) -> Self {
        Self {
            player_id: value.player_id,
            balance: value.balance,
        }
    }
}

#[derive(Default, BorshDeserialize, BorshSerialize, Debug, PartialEq, Eq, Clone)]
pub enum GameStatus {
    #[default]
    Initializing,
    Initialized,
    Closed,
}

// State of on-chain GameAccount
#[cfg_attr(test, derive(PartialEq, Clone))]
#[derive(Default, BorshDeserialize, BorshSerialize, Debug)]
pub struct GameState {
    pub game_status: GameStatus,
    // the contract version, used for upgrade
    pub version: String,
    // game name displayed on chain
    pub title: String,
    // addr to the game core logic program on Arweave
    pub bundle_addr: Pubkey,
    // addr to the account that holds all players' deposits
    pub stake_account: Pubkey,
    // game owner who created this game account
    pub owner: Pubkey,
    // mint id of the token used for game
    pub token_mint: Pubkey,
    // addr of the first server joined the game
    pub transactor_addr: Option<Pubkey>,
    // a serial number, increased by 1 after each PlayerJoin or ServerJoin
    pub access_version: u64,
    // a serial number, increased by 1 after each settlement
    pub settle_version: u64,
    // game size
    pub max_players: u16,
    // the address to players reg account
    pub players_reg_account: Pubkey,
    // player deposits
    pub deposits: Vec<PlayerDeposit>,
    // game servers (max: 10)
    pub servers: Vec<ServerJoin>,
    // length of game-specific data
    pub data_len: u32,
    // serialized data of game-specific data such as sb/bb in Texas Holdem
    pub data: Vec<u8>,
    // game votes
    pub votes: Vec<Vote>,
    // unlock time
    pub unlock_time: Option<u64>,
    // the entry type
    pub entry_type: EntryType,
    // the recipient account
    pub recipient_addr: Pubkey,
    // the checkpoint state
    pub checkpoint: Vec<u8>,
    // the lock for entry
    pub entry_lock: EntryLock,
    // a list of bonuses that can be awarded in game
    pub bonuses: Vec<Bonus>,
    // the snapshot for checkpoint balances
    pub balances: Vec<PlayerBalance>,
}

impl GameState {
    pub fn into_account<S: Into<String>>(self, addr: S, players: Vec<PlayerJoin>) -> Result<GameAccount, Error> {
        let GameState {
            title,
            bundle_addr,
            owner,
            token_mint,
            transactor_addr,
            access_version,
            settle_version,
            max_players,
            servers,
            data_len,
            data,
            entry_type,
            recipient_addr,
            checkpoint,
            entry_lock,
            deposits,
            bonuses,
            balances,
            ..
        } = self;

        let players = players.into_iter().map(Into::into).collect();
        let servers = servers.into_iter().map(Into::into).collect();
        let deposits = deposits.into_iter().map(Into::into).collect();
        let bonuses = bonuses.into_iter().map(Into::into).collect();
        let balances = balances.into_iter().map(Into::into).collect();

        let checkpoint_onchain = if !checkpoint.is_empty() {
            Some(
                CheckpointOnChain::try_from_slice(&checkpoint)
                    .map_err(|_| Error::MalformedCheckpoint)?,
            )
        } else {
            None
        };

        Ok(GameAccount {
            addr: addr.into(),
            title,
            settle_version,
            bundle_addr: bundle_addr.to_string(),
            token_addr: token_mint.to_string(),
            owner_addr: owner.to_string(),
            access_version,
            players,
            servers,
            transactor_addr: transactor_addr.map(|pk| pk.to_string()),
            max_players,
            data_len,
            data,
            deposits,
            votes: Vec::new(),
            unlock_time: None,
            recipient_addr: recipient_addr.to_string(),
            entry_type,
            checkpoint_on_chain: checkpoint_onchain,
            entry_lock,
            bonuses,
            balances,
        })
    }
}
